python/cudf/cudf/_lib/spillable_buffer.pyx

```diff
@@ -0,0 +1,448 @@
+                self._owner = None
+            elif (ptr_type, target) == ("cpu", "gpu"):
+                # Notice, this operation is prone to deadlock because the RMM
+                # allocation might trigger spilling-on-demand which in turn
+                # trigger a new call to this buffer's `__spill__()`.
+                # Therefore, it is important that spilling-on-demand doesn't
+                # tries to unspill an already locked buffer!
+                dev_mem = rmm.DeviceBuffer.to_device(
+                    self._ptr_desc.pop("memoryview")
+                )
+                self._ptr = dev_mem.ptr
+                self._size = dev_mem.size
+                self._owner = dev_mem
+            else:
+                # TODO: support moving to disk
+                raise ValueError(f"Unknown target: {target}")
+            self._ptr_desc["type"] = target
+
+    @property
+    def ptr(self) -> int:
+        """Access the memory directly
+
+        Notice, this will mark the buffer as "exposed" and make
+        it unspillable permanently.
+
+        Consider using `ptr_raw() or `.restricted_ptr()` instead.
+        """
+        if self._view_desc:
+            return self._view_desc["base"].ptr + self._view_desc["offset"]
+        self._manager.spill_to_device_limit()
+        with self._lock:
+            if not self._exposed:
+                self._manager.log_expose(self)
+            self.__spill__(target="gpu")
+            self._exposed = True
+            self._last_accessed = time.monotonic()
+            return self._ptr
+
+    cdef void* ptr_raw(self, SpillLock spill_lock) except *:
+        if spill_lock is None:
+            return <void*><uintptr_t> self.ptr
+
+        # Get base buffer
+        cdef SpillableBuffer base
+        cdef size_t offset
+        if self._view_desc is None:
+            base = self
+            offset = 0
+        else:
+            base = self._view_desc["base"]
+            offset = self._view_desc["offset"]
+
+        with base._lock:
+            base.__spill__(target="gpu")
+            base._last_accessed = time.monotonic()
+            spill_lock.add(base._expose_counter)
+            return <void*><uintptr_t> (base._ptr+offset)
+
+    def ptr_restricted(self) -> Tuple[int, SpillLock]:
+        # Get base buffer
+        cdef SpillableBuffer base
+        cdef size_t offset
+        if self._view_desc is None:
+            base = self
+            offset = 0
+        else:
+            base = self._view_desc["base"]
+            offset = self._view_desc["offset"]
+
+        with base._lock:
+            base.__spill__(target="gpu")
+            base._last_accessed = time.monotonic()
+            spill_lock = SpillLock()
+            spill_lock.add(base._expose_counter)
+            return base._ptr+offset, spill_lock
+
+    @property
+    def owner(self) -> Any:
+        return self._owner
+
+    @property
+    def exposed(self) -> bool:
+        if self._view_desc:
```

Comment:
I think it would be worth defining a decorator to capture the pattern `if self._view_desc: call method on self._view_desc["base"]`.

Reply:
Removed `self._view_desc`, we now have a `class SpillableBufferView(SpillableBuffer)`